controllers/teleport_controller/teleport_controller.cpp

```diff
@@ -0,0 +1,138 @@
+// File:          teleport_controller.cpp
+// Date:          23/03/2021
+// Description:   Method to teleport robot around the playing field for the purposes of
+//                    gathering training data
+// Author:        Lachlan Court
+// Modifications:
+
+
+#include <RobotisOp2GaitManager.hpp>
+#include <RobotisOp2MotionManager.hpp>
+#include <webots/Supervisor.hpp>
+#include <iostream>
+#include <math.h>
+#include <time.h> 
+
+// All the webots classes are defined in the "webots" namespace
+using namespace webots;
+
+/*
+// Rotations matrix for 90 degree rotations
+static const int rotationsSize = 4;
+static const double rotations[rotationsSize][4] = {{1, 0, 0, 1.57081},
+                                                  {0.577353, 0.577345, 0.577353, 2.0944},
+                                                  {1.69526e-09, 0.707103, 0.707111, 3.14159},
+                                                  {-0.577352, 0.577347, 0.577352, -2.09441}};
+
+*/
+
+// Rotations matrix for 45 degree rotations
+static const int rotationsSize = 8;                                     
+static const double rotations[rotationsSize][4] = {{1, 0, 0, 1.57081},
+                                                  {0.867737, 0.343283, 0.359429, 1.75753},
+                                                  {0.577353, 0.577345, 0.577353, 2.0944},
+                                                  {0.286947, 0.661632, 0.692752, 2.60662},
+                                                  {1.69526e-09, 0.707103, 0.707111, 3.14159},
+                                                  {-0.286948, 0.661631, 0.692752, -2.60661},
+                                                  {-0.577352, 0.577347, 0.577352, -2.09441},
+                                                  {0.867739, -0.343277, -0.359429, 1.75753}};                               
+
+
+/* This is the main program of the supervisor. It creates an instance of the Supervisor
+ *     instance, launches its function(s) and destroys it at the end of the execution.
+ */
+int main() {
+    std::cout << "a" << std::endl;
+    // create the Supervisor instance and assign it to a robot
+    Supervisor* supervisor = new Supervisor();
+    Node* target           = supervisor->getFromDef("RED_1");
+    std::cout << "b" << std::endl;
+    // Get the time step of the current world.
+    int timeStep = (int) supervisor->getBasicTimeStep();
```

Comment:
Never use c-style casts. Use functional casts, static casts, or even reinterpret casts (reinterpret casts are usually ok for our purposes, despite what stackoverflow says, because we aren't doing security critical things or anything redhot which makes reinterpret casts a hot topic. @TrentHouliston has many more things to say about such things lol).

Reply:
Not sure what this means sorry, that's the only way I've ever done a typecast

Reply:
[Relevant cppreference page](https://en.cppreference.com/w/cpp/language/explicit_cast). Typecasts can be a bit of a rabbit hole, but it's interesting enough. 

Here are some heuristics for casting in our codebase:

- `(type) thing` is a c-style cast. These are bad
- `type(thing)` is a functional cast. these are fine
- `static_cast<type>(thing)` is a static cast. these are fine
- `reinterpret_cast<type>(thing)` is a reinterpret cast. People online will yell at you for using these, but not for reasons we care about in our code

Reply:
Yeah, a c style cast is a sledgehammer it will do a static cast and a reinterpret cast at the same time. This is bad because we really don't want to be using reinterpret casts by accident.

Also, it's harder with a c style cast to know what the cast is applying to
`int(a) + b` is much easier to understand that the a is getting cast to an int rather than `(int) a + b` where it's the same, but harder to tell that it's a that's cast, and not `a + b`

Reply:
Wait int(a) works in c++?? Ahhhhhhh
I do that in python but I didn't realise you could do in c++ as well 😭

Reply:
I mean technically what you're doing is constructing a new `int` instance using the value you pass in as constructor arguments. But it basically works as a cast

Reply:
Ohhhhh okay okay neat that makes sense